### PR TITLE
Docs/ Add AGENTS.md

### PR DESCRIPTION
**Pull request by PetromirDev**
## Changes:
- Adds a brief root `AGENTS.md` file
- Adds an `AGENTS.md` file for controllers with specific information

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+You are an AI Agent working in the core business-logic repository for the Ambire wallet. This code is used by multiple environments (mobile, extension, standalone websites) and is a critical part of the wallet's security and functionality. Changes to this code can have wide-reaching implications across all platforms, so extreme caution is required when making updates.
```
Comment (superKalo) [inline AGENTS.md]: I'd add after `Ambire wallet` this - `, a self-custodial Web3 wallet for managing digital assets, smart accounts, DeFi interactions, and blockchain transactions across multiple EVM-compatible networks.`

Also - `(mobile, extension, standalone websites)` - I'd change to "(iOS and Android mobile apps, browser extensions for Chrome and Firefox, web apps).

File: AGENTS.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+## Tech stack
+typescript, ethers, viem, jest, hardhat
```
Comment (superKalo) [inline AGENTS.md:4]: Some I saw also have "common commands" section. I wonder if we should add:

```
## Common commands
- Install dependencies: `npm install`
- Run one Jest test file: `npm run jest -- path/to/file.test.ts`
- Type-check: `npm run type:check`
- Lint and auto-fix source files: `npm run lint:fix`
- Compile contracts: `npm run compile:contracts`
- Run Hardhat compile + tests: `npm run hardhat`
```

(for the last two we need to checkbase with Yosif / Bobby).

File: AGENTS.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+### Code quality:
+- Always ensure subscriptions, event listeners, timers and other side effects are properly cleaned up
+- NEVER delete existing comments when updating a code block; update inaccurate comments instead. Delete a comment only if the logic it describes is completely removed or the new logic is entirely self-explanatory
+- NEVER swallow errors; log them and handle appropriately.
+- NEVER modify git config or run destructive git operations
+- NEVER commit unless explicitly requested by user
+- NEVER stage changes unless explicitly requested by user
+
+## Tests:
+- ALWAYS write/update existing tests for changes that you make. Propose missing test cases if you find gaps in the existing coverage.
```
Comment (superKalo) [inline AGENTS.md]: I'm not against this, but it may bring a lot of overhead for simpler / smaller changes here and there. I wonder if we should instead say "for changes that you make" tweak it to be "for behavior changes", wdyt?

File: AGENTS.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+- NEVER delete existing comments when updating a code block; update inaccurate comments instead. Delete a comment only if the logic it describes is completely removed or the new logic is entirely self-explanatory
+- NEVER swallow errors; log them and handle appropriately.
+- NEVER modify git config or run destructive git operations
+- NEVER commit unless explicitly requested by user
+- NEVER stage changes unless explicitly requested by user
+
+## Tests:
+- ALWAYS write/update existing tests for changes that you make. Propose missing test cases if you find gaps in the existing coverage. 
+- ALWAYS write test cases that cover positive, negative, edge cases and security implications of the code you change or add.
+- ALWAYS make a quick tool call to another controller/library when writing new code to check how tests are written there and follow the same patterns. NEVER invent new testing patterns before checking existing ones in the codebase.
```
Comment (superKalo) [inline AGENTS.md]: Could be shortened a little to something like: "Before adding a new testing pattern, inspect nearby `*.test.ts` files or tests for a similar controller/library and follow the local style."

File: AGENTS.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+- `src/libs/` contains pure business logic without side effects
+- `src/contracts/` contains Solidity contracts used by the wallet and relayer; compiled artifacts live in `contracts/compiled/`
+- `src/services/` contains external service integrations (bundlers, RPC, ENS, etc.)
+
+## Rules
+
+### Repository specific:
+- Use `npm` for package management
+- All code must be environment-agnostic; do not use platform-specific APIs (e.g. `window`, `document`, `React Native` APIs)
+- NO new RPC providers should be constructed. ALWAYS use `ProviderController`.
```
Comment (superKalo) [inline AGENTS.md]: Controller name is: Provider**s**Controller

File: AGENTS.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
+You are an AI Agent working in the core business-logic repository for the Ambire wallet. This code is used by multiple environments (mobile, extension, standalone websites) and is a critical part of the wallet's security and functionality. Changes to this code can have wide-reaching implications across all platforms, so extreme caution is required when making updates.
+
+## Tech stack
+typescript, ethers, viem, jest, hardhat
+
+## Project overview
```
Comment (superKalo) [inline AGENTS.md:6]: Maybe we should mention here or in another section that "This package is not intended to run standalone.", wdyt? Maybe in the beginning.

File: AGENTS.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+## Rules
+
+### Repository specific:
+- Use `npm` for package management
+- All code must be environment-agnostic; do not use platform-specific APIs (e.g. `window`, `document`, `React Native` APIs)
+- NO new RPC providers should be constructed. ALWAYS use `ProviderController`.
+
+### Code quality:
+- Always ensure subscriptions, event listeners, timers and other side effects are properly cleaned up
```
Comment (superKalo) [inline AGENTS.md:32]: Also maybe we should add "Code usually runs continuously for long periods of time (especially in browser extension environments), so memory leaks, listeners, and orphaned async processes can accumulate and cause performance, stability, or reliability issues over time."

File: AGENTS.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+- NEVER delete existing comments when updating a code block; update inaccurate comments instead. Delete a comment only if the logic it describes is completely removed or the new logic is entirely self-explanatory
```
Comment (superKalo) [inline AGENTS.md:33]: 🔝 , AI ofter deletes precious comments on my end when does refactoring!

File: AGENTS.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
+You are an AI Agent working in the core business-logic repository for the Ambire wallet. This code is used by multiple environments (mobile, extension, standalone websites) and is a critical part of the wallet's security and functionality. Changes to this code can have wide-reaching implications across all platforms, so extreme caution is required when making updates.
+
+## Tech stack
+typescript, ethers, viem, jest, hardhat
+
+## Project overview
+- `src/controllers/` contains stateful classes that extend `EventEmitter`; they propagate updates via `this.emitUpdate()`
+- `src/libs/` contains pure business logic without side effects
+- `src/contracts/` contains Solidity contracts used by the wallet and relayer; compiled artifacts live in `contracts/compiled/`
+- `src/services/` contains external service integrations (bundlers, RPC, ENS, etc.)
```
Comment (superKalo) [inline AGENTS.md:11]: We should add this chaveat:

- This package **does not** include compiled **JS and TS** output in the repository. Each Ambire app compiles it individually as needed. The `dist/` folder exists for internal use only. It may contain some compiled files, but it is not kept up to date and not updated on release. Do not rely on dist/ as a source of compiled code.

File: src/controllers/AGENTS.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+- Long-running background intervals must be declared in `ContinuousUpdatesController`, which orchestrates their lifecycle based on app state and controller events. If you need a new background loop, add it there and wire its start/stop/restart logic through the existing event subscriptions.
+- Never call `this.storage.set()` in parallel. Always await the previous call before making another one.
+- Always use `this.emitError` for error handling in controllers; all emitted errors are reported to Sentry and logged, and non-silent errors are also displayed as toasts in the UI. Public methods must never let errors propagate — use `EmittableError` (thrown inside a `withStatus` wrapper, which auto-emits it) or `try/catch` + `emitError({ level, message, error })` otherwise.
+- Public state is serialized and sent to the UI on every update, so it should be minimal and only include what's necessary for the UI. Do not store large data or sensitive data in public state. Use private fields for that and expose only derived non-sensitive data in public state if needed.
+- NEVER write expensive calculations inside getters.
```
Comment (superKalo) [inline src/controllers/AGENTS.md:85]: and also:

- in some occasions avoid doing async stuff, right
- do not emitUpdates in getters (infinite loops)

EDIT: I'm now looking at the rule below and I think it prob summarizes some of these conserns

File: src/controllers/AGENTS.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+- If a controller depends on the state of the UI (e.g., which screen it is on), it should subscribe to `this.ui.uiEvent.on`
+- When retrying failed background fetches, use a retry counter with a maximum number of attempts (reset on success) and an increasing delay. For periodic polling with retry, use `RecurringTimeout` with adaptive intervals (shorter on failure, longer on success). See `PortfolioController.updateExchangeList()`, `DappsController.#retryFetchAndUpdateInterval`, and `ContractNamesController`'s `retryAfter` timestamps for examples.
+- ALWAYS guard async operations that update state with appropriate stale-data checks, such as debounce, unique ID/version checks, or cancellation with `AbortController`, to prevent
+state corruption from out-of-order or concurrent operations. Examples of these patterns can be found in `SwapAndBridgeController` and `AccountPickerController`.
```
Comment (superKalo) [inline src/controllers/AGENTS.md]: Mega minor: there should be no new line here 👮

File: src/controllers/AGENTS.md [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+- **ContinuousUpdatesController** – Orchestrates periodic background updates for multiple controllers (e.g., portfolio and activity)
+- **ContractNamesController** – Resolves human-readable names for smart-contract addresses via the relayer.
+- **DappsController** – Manages dApp connections, sessions, verification status, and the dApp catalog.
+- **DomainsController** – Resolves and caches ENS and Namoshi names (and avatars) for addresses.
+- **EmailVaultController** – Handles email-based recovery, magic-link flows, and vault secret management.
+- **FeatureFlagsController** – Toggles application features at runtime for roll-outs and A/B testing.
+- **EstimationController** – Estimates gas, fees, and payment options for smart-account transactions.
+- **GasPriceController** – Fetches and formats gas-price recommendations and bundler gas speeds.
+- **InviteController** – Manages invite codes and OG status (legacy; now used for status tracking only).
+- **KeystoreController** – Encrypts, decrypts, and manages private keys, seeds, and key preferences.
```
Comment (superKalo) [inline src/controllers/AGENTS.md]: I'd teak this to: Encrypts seeds and private keys **under a multi-secret–wrapped main key**, manages unlock state, and routes signing to internal or hardware-backed keys.

**Review by superKalo — APPROVED** [review]
Great work summarizing the key agent instructions clearly and keeping the guidance focused 👏 👏 👏 

I added some notes, but in general, I like it a lot.

My side concern is how we should receive feedback that something is working great or terrible after these instructions went in? Maybe we should communicate with all colleagues what they should stay alert for. Let's sync offline for this.

File: AGENTS.md (hunk 3 above)
Comment (PetromirDev) [reply AGENTS.md]: Let's remove this
